libp2p-kad/src/kbucket.rs

```diff
@@ -0,0 +1,474 @@
+                    pending_node: None,
+                    last_update: Instant::now(),
+                })
+                .map(Mutex::new)
+                .collect(),
+            ping_timeout: ping_timeout,
+        }
+    }
+
+    // Returns the id of the bucket that should contain the peer with the given ID.
+    //
+    // Returns `None` if out of range, which happens if `id` is the same as the local peer id.
+    #[inline]
+    fn bucket_num(&self, id: &Id) -> Option<usize> {
+        (Id::num_bits() - 1).checked_sub(Id::leading_zeros(self.my_id.distance_with(id)) as usize)
+    }
+
+    /// Returns an iterator to all the buckets of this table.
+    ///
+    /// Ordered by proximity to the local node. Closest bucket (with max. one node in it) comes
+    /// first.
+    #[inline]
+    pub fn buckets(&self) -> BucketsIter<Id, Val> {
+        BucketsIter(self.tables.iter(), self.ping_timeout)
+    }
+
+    /// Returns the ID of the local node.
+    #[inline]
+    pub fn my_id(&self) -> &Id {
+        &self.my_id
+    }
+
+    /// Finds the `num` nodes closest to `id`, ordered by distance.
+    pub fn find_closest(&self, id: &Id) -> VecIntoIter<Id>
+    where
+        Id: Clone,
+    {
+        // TODO: optimize
+        let mut out = Vec::new();
+        for table in self.tables.iter() {
+            let mut table = table.lock();
+            table.flush(self.ping_timeout);
+            for node in table.nodes.iter() {
+                out.push(node.id.clone());
+            }
+        }
+        out.sort_by(|a, b| b.distance_with(id).cmp(&a.distance_with(id)));
+        out.into_iter()
+    }
+
+    /// Marks the node as "most recent" in its bucket and modifies the value associated to it.
+    /// This function should be called whenever we receive a communication from a node.
+    ///
+    /// # Panic
+    ///
+    /// Panics if `id` is equal to the local node ID.
+    ///
+    pub fn update(&self, id: Id, value: Val) -> UpdateOutcome<Id, Val> {
+        let table = match self.bucket_num(&id) {
+            Some(n) => &self.tables[n],
+            None => panic!("tried to update our own node in the kbuckets table"),
+        };
+
+        let mut table = table.lock();
+        table.flush(self.ping_timeout);
+
+        if let Some(pos) = table.nodes.iter().position(|n| n.id == id) {
+            // Node is already in the bucket.
+            let mut existing = table.nodes.remove(pos);
+            let old_val = mem::replace(&mut existing.value, value);
+            if pos == 0 {
+                // If it's the first node of the bucket that we update, then we drop the node that
+                // was waiting for a ping.
+                table.nodes.truncate(MAX_NODES_PER_BUCKET - 1);
+                table.pending_node = None;
+            }
+            table.nodes.push(existing);
+            table.last_update = Instant::now();
+            UpdateOutcome::Refreshed(old_val)
+        } else if table.nodes.len() < MAX_NODES_PER_BUCKET {
```

Comment:
If you're trying to clamp space use, use [`arrayvec`](https://docs.rs/arrayvec/0.4.7/arrayvec/) and [`ArrayVec::try_push`](https://docs.rs/arrayvec/0.4.7/arrayvec/struct.ArrayVec.html#method.try_push). It compiles _way_ better and is much easier to read.